internal/appsec/dyngo/operation.go

```diff
@@ -34,50 +33,23 @@ import (
 // operation once it finishes so that it no longer can be called on finished
 // operations.
 type Operation interface {
-	// On allows to register an event listener to the operation. The event
-	// listener will be removed from the operation once it finishes.
-	On(EventListener)
-
-	// OnData allows to register a data listener to the operation
-	OnData(DataListener)
-
-	// EmitData sends data to the data listeners of the operation
-	EmitData(any)
-
-	// Parent return the parent operation. It returns nil for the root
-	// operation.
+	// Parent returns the parent operation, or nil for the root operation.
 	Parent() Operation
 
-	// emitEvent emits the event to listeners of the given argsType and calls
-	// them with the given op and v values.
-	// emitEvent is a private method implemented by the operation struct type so
-	// that no other package can define it.
-	emitEvent(argsType reflect.Type, op Operation, v interface{})
-
-	emitData(argsType reflect.Type, v any)
-
-	// add the given event listeners to the operation.
-	// add is a private method implemented by the operation struct type so
-	// that no other package can define it.
-	add(...EventListener)
+	// self is an internal method guaranteeing only *operation implements Operation.
+	self() *operation
```

Comment:
Nice simplification indeed 👏
But while I was reading its use in the code, I was thinking `o.implementation()` or `o.unwrap()` might be more understandable maybe:
```suggestion
	 unwrap() *operation
```

No strong opinion here.